File: longestSubstring.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        int windowStart = 0, windowEnd = 0, n = s.length(), max = 0, size = 0;
        
        unordered_map<char, int> map;
        
        for(windowEnd = 0; windowEnd < n; windowEnd++)
        {
            map[s[windowEnd]]++;
            
            while(map[s[windowEnd]] > 1){
                
                map[s[windowStart]]--;
                
                windowStart++;
            }
            
            size = windowEnd - windowStart + 1;
            
            if(size > max) max = size;
        }
        
        return max;
        
    }
};

File: progress_sheet/longestSubstring.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        left = 0
        size = len(s)
        visited = set()
        longest = 0

        for right in range(size):

            while s[right] in visited:
                visited.remove(s[left])
                left += 1

            visited.add(s[right])
            longest = max(longest, right - left + 1)
        
        return longest
